arch/x86/core/userspace.S

```diff
@@ -0,0 +1,145 @@
+ * further processing.
+ */
+SECTION_FUNC(TEXT, _x86_syscall_entry_stub)
+	push	%esi		/* arg5 */
+	push	%edi		/* arg4 */
+	push	%ebx		/* arg3 */
+#ifndef CONFIG_X86_IAMCU
+	push	%ecx		/* arg2 */
+	push	%edx		/* arg1	*/
+	push	%eax		/* Call id */
+#endif
+
+	call	_x86_syscall_entry
+
+	/* EAX now contains return value. Pop or xor everything else to prevent
+	 * information leak from kernel mode.
+	 */
+#ifndef CONFIG_X86_IAMCU
+	add	$4, %esp	/* old EAX value */
+	pop	%edx
+	pop	%ecx
+#else
+	xor	%edx, %edx
+	xor	%ecx, %ecx
+#endif
+	pop	%ebx
+	pop	%edi
+	pop	%esi
+
+	iret
+
+
+/* FUNC_NORETURN void _x86_userspace_enter(_thread_entry_t user_entry,
+ *					   void *p1, void *p2, void *p3,
+ *					   u32_t stack_end,
+ *					   u32_t stack_start)
+ *
+ * A one-way trip to userspace.
+ */
+SECTION_FUNC(TEXT, _x86_userspace_enter)
+	add	$0x4, %esp /* Discard return address on stack */
+
+	/* Fetch parameters on the stack */
+#ifndef CONFIG_X86_IAMCU
+	pop	%eax	/* user_entry */
+	pop	%edx	/* p1 */
+	pop	%ecx	/* p2 */
+#endif
+	pop	%esi	/* p3 */
+	pop	%ebx	/* stack_end (high address) */
+	pop	%edi	/* stack_start (low address) */
+
+	/* Move to the kernel stack for this thread, so we can erase the
+	 * user stack. The kernel stack is the page immediately before
+	 * the user stack.
+	 *
+	 * For security reasons, we must erase the entire user stack.
+	 * We don't know what previous contexts it was used and do not
+	 * want to leak any information.
+	 */
+	mov	%edi, %esp
+
+	/* Stash some registers we are going to need to erase the user
+	 * stack.
+	 */
+	push	%ecx
+	push	%edi
+	push	%eax
+
+	/* Compute size of user stack and put in ECX */
+	mov	%ebx, %ecx
+	sub	%edi, %ecx
+
+#ifdef CONFIG_INIT_STACKS
+	mov	$0xAAAAAAAA, %eax
+#else
+	xor	%eax, %eax
+#endif
+	/* Fill ECX bytes of memory, 4 bytes at a time, starting at EDI,
+	 * with whatever is in EAX. Stack sizes are always at least 4-byte
+	 * aligned
+	 */
+	cld
+	rep
+	stosl
+
+	/* Restore registers */
+	pop	%eax
+	pop	%edi
+	pop	%ecx
+
+	/* Now set stack pointer to the base of the user stack. Now that this
+	 * is set we won't need EBX any more.
+	 */
+	mov	%ebx, %esp
+
+	/* Set segment registers (except CS and SS which are done in
+	 * a special way by 'iret' below)
+	 */
+	mov	$USER_DATA_SEG, %bx
+	mov	%bx, %ds
+	mov	%bx, %es
+	mov	%bx, %fs
+	mov	%bx, %gs
+
+	/* Push arguments to _thread_entry() */
+	push	%esi	/* p3 */
+#ifndef CONFIG_X86_IAMCU
+	push	%ecx	/* p2 */
+	push	%edx	/* p1 */
+	push	%eax	/* user_entry */
+#endif
+	/* NULL return address */
+	push	$0
+
+	/* Save stack pointer at this position, this is where it will be
+	 * when we land in _thread_entry()
+	 */
+	mov	%esp, %edi
+
+	/* Inter-privilege 'iret' pop all of these */
+	push	$USER_DATA_SEG	/* SS */
+	push	%edi		/* ESP */
+	pushfl			/* EFLAGS */
+	push	$USER_CODE_SEG	/* CS */
+	push	$_thread_entry	/* EIP */
+
+	/* We will land in _thread_entry() in user mode after this */
+	iret
```

Comment:
This seems a little clumsy, having to fake an interrupt stack frame.  Can't you just do a far call here?  Must the initial entry really be an IRET?

Reply:
to make a long story short, no. it has to be an iret.